[bootstrap-vcpkg.sh] Test for ${CC} rather than gcc on systems using vcpkgUseSystem

SchaichAlonso opened the pull request:
Fixes #36807

SchaichAlonso committed:
Properly test for variable definition

SchaichAlonso committed:
If alpine uses "gcc", OpenBSD should use "clang" rather then "/usr/bin/clang"

SchaichAlonso committed:
Check for ${CC} rather then gcc

Check for `${CC}` and `${CXX}` to be installed rather then checking for `gcc`.

On Alpine, we check for `gcc` to be present, so setup `${CC}` and `${CXX}` accordingly in order to actually use the gcc that has been asserted to exist. This is nontrivial because [around line 201](https://github.com/microsoft/vcpkg/blob/13bde2ff13192e1b2fdd37bd9b475c7665ae6ae5/scripts/bootstrap.sh#L201), ${CXX}` might be set to a value that is not related to the `gcc` (distribution's default versioned) package.

SchaichAlonso committed:
Use default value assignments for CC/CXX

dg0yt commented on scripts/bootstrap.sh:
```diff
@@ -98,24 +98,22 @@ if [ -e /etc/alpine-release ]; then
     vcpkgCheckRepoTool ninja
     vcpkgCheckRepoTool git
-    vcpkgCheckRepoTool gcc
+    vcpkgCheckRepoTool ${CC}
+    vcpkgCheckRepoTool ${CXX}
```
I think this is to early.
There is the legacy gcc cascade starting at line 196. Legacy because it should be remove for simply using gcc and g++.
After this choice, there should be the vcpkgCheckRepoTool for CC/CXX.

And BTW can't we simply use `cc` / `c++`, instead of favoring particular vendors?

SchaichAlonso replied:
From a logical perspective, it is too early. Alphine Linux systems require `gcc` to be installed but might later on decide to use `gcc-12` instead, making the `gcc` requirement obsolete.

Using cc and c++ as default values for `CC` and `CXX` is probably more straightforward then manually enforcing gcc or clang. Users that aren't happy with their system's default compiler probably export `CC`/`CXX` system-wide anyways.

The version selection for gcc is quite fishy, anyways. Distributors deciding to install gcc with some suffixes usually also patch their cmake package to detect the suffixed gcc, so the whole cascade can be dropped as cmake will do the right thing anyways, while our manual preference setup lacks a branch for gcc-13 which "most likely" should be prefered over gcc-6.

dg0yt replied:
> The version selection for gcc is quite fishy, anyways. 

This was discussed before, and there is support for removing it. Just nobofy did it...

SchaichAlonso replied:
I'll remove it then

SchaichAlonso committed:
Require system default compilers rather then gcc/clang

SchaichAlonso committed:
Have cmake decide on a compiler if CXX is not provided

Distributors that deploy custom-named compilers also patch their cmake deployments to detect these compilers.

If multiple compilers are available on a system, have cmake decide which one to use. CMake will automatically try to use `${CXX}` if the variable is set, and otherwise decide on a compiler in a the distribution/user configurable way.

SchaichAlonso committed:
Do not run an explicit compiler presence check when vcpkgUseSystem is ON

cmake will fail if it cannot find a compiler, and in consequence, bootstrap-vcpkg will fail.

Remove the explicit check for system compiler presence on Alpine and OpenBSD, and have cmake implicitly decide on a compiler instead.

dg0yt reviewed (approved):
LGTM.
Let's see what vcpkg team says.

jimwang118 commented:
The test script runs correctly on Linux. Add reviewed tag for further review by the vcpkg team.

BillyONeal commented on scripts/bootstrap.sh:
```diff
@@ -197,27 +189,6 @@ fi
 if [ "$vcpkgDownloadTool" = "ON" ]; then
     vcpkgDownloadFile "https://github.com/microsoft/vcpkg-tool/releases/download/$VCPKG_TOOL_RELEASE_TAG/$vcpkgToolName" "$vcpkgRootDir/vcpkg" $vcpkgToolReleaseSha
 else
-    if [ "x$CXX" = "x" ]; then
```
For what it's worth, I believe this stuff was here to deal with when we needed `std::filesystem` and the compilers on the other platforms. I think on the platforms where this was needed we now provide binaries so I'm happy with removing it; but strictly speaking it seems unrelated to the slated issue...
